src/common/ceph_context.h

```diff
@@ -79,11 +79,42 @@ class CephContext {
   crimson::common::PerfCountersCollection& _perf_counters_collection;
   CephContext* get();
   void put();
+
+  /**
+   * Get the admin socket associated with this CephContext.
+   *
+   * Note that an admin socket should only be created for
+   * daemons.
+   *
+   * @return the admin socket or nullptr
+   */
+  crimson::admin::AdminSocket* get_admin_socket() { 
+    if (asok) {
+      return asok.get();
+    }
+    return nullptr;
+  }
+
+  void release_admin_socket();
+
+  crimson::admin::ContextConfigAdmin* get_config_admin() { 
+    if (asok_config_admin) {
+      return asok_config_admin.get();
+    }
+    return nullptr;
+  }
+
 private:
   std::unique_ptr<CryptoRandom> _crypto_random;
   unsigned nref;
+  seastar::lw_shared_ptr<crimson::admin::AdminSocket> asok;
```

Comment:
why do you want to make the asock service a member variable of CephContext? i think it should be a mini service resided in OSD.

Reply:
I really don't, and I agree it's the wrong place for it. But that's how classic OSD was structured.
Note that some of the asok commands are to be handled in the context of the CephContext, so there will be an 'asok element' there anyway.
I'd suggest merging the PR with the original design for now, then thinking whether we want to stray from the original design on this point.

Reply:
> But that's how classic OSD was structured.

i don't think we are trying to replicate the design of OSD.

> I'd suggest merging the PR with the original design for now, then thinking whether we want to stray from the original design on this point.

i am inclined to fix the design in crimson-osd if we are not reusing the existing OSD facilities.

Reply:
fixed